[iOS] Cannot scroll page by dragging on a bare <video> element

jernoble opened the pull request:
#### 7fa44a650d9a3604421d1c3807e48a0feaa85706
<pre>
[iOS] Cannot scroll page by dragging on a bare &lt;video&gt; element
<a href="https://bugs.webkit.org/show_bug.cgi?id=261563">https://bugs.webkit.org/show_bug.cgi?id=261563</a>
rdar://114720841

Reviewed by Eric Carlson.

Hit testing in the WebKit utility method isScrolledBy() works by walking up the view heirarchy looking for a view which return a valid node from RemoteLayerTreeNode::forCALayer(). However, while WKLayerHostView is a subview of WebAVPlayerLayerView, it&apos;s superview property is nil. This occurs because while the WKLayerHostView.layer is made a sublayer of the WebAVPlayerLayerView.playerLayer, the WKLayerHostView itself is not made a subview of the WebAVPlayerLayerView.

Drive-by fix: in testing this patch, a crash will occasionally occur where a null pointer is dereferenced in collectDescendantViewsAtPoint(). Add a null check and return false if that check fails.

* LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow.html: Added.
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::collectDescendantViewsAtPoint):

Canonical link: <a href="https://commits.webkit.org/268566@main">https://commits.webkit.org/268566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82cb168e0a8a5607ac27d257d124dff343d7f27b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21113 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21955 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18722 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20653 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22805 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17371 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18228 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24482 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18404 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/22472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18997 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16116 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18064 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4806 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18823 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow.html:
```diff
@@ -0,0 +1,41 @@
+
+        let middleX = video.offsetLeft + video.offsetWidth / 2;
+        let middleY = video.offsetTop + video.offsetHeight / 2;
+        touchAndDragFromPointToPoint(middleX, middleY, middleX, 0);
```
You need to `await` a single function which calls `touchAndDragFromPointToPoint` and `liftUpAtPoint` sequentially

smfr commented on LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow.html:
```diff
@@ -0,0 +1,41 @@
+    <style>
+        html, body { height: 100vh; overflow: hidden; }
+        #spacer { width: 100vw; height: 120vh; }
+        #scroller { border: 1px solid red; height: 50vh; width: 100vw; overflow-y: scroll; }
```
No red in a passing test please.

jernoble committed:
[iOS] Cannot scroll page by dragging on a bare <video> element
https://bugs.webkit.org/show_bug.cgi?id=261563
rdar://114720841

Reviewed by Eric Carlson.

Hit testing in the WebKit utility method isScrolledBy() works by walking up the view heirarchy looking for a view which return a valid node from RemoteLayerTreeNode::forCALayer(). However, while WKLayerHostView is a subview of WebAVPlayerLayerView, it's superview property is nil. This occurs because while the WKLayerHostView.layer is made a sublayer of the WebAVPlayerLayerView.playerLayer, the WKLayerHostView itself is not made a subview of the WebAVPlayerLayerView.

Drive-by fix: in testing this patch, a crash will occasionally occur where a null pointer is dereferenced in collectDescendantViewsAtPoint(). Add a null check and return false if that check fails.

* LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/video-scrolling-with-overflow.html: Added.
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::collectDescendantViewsAtPoint):

Canonical link: https://commits.webkit.org/268566@main
